### ISSUE TITLE
Properly handle collisions inside bubbles (for boids)

### ISSUE THREAD

**Issue by JVillella** [huawei-noah/SMARTS#18]
When a collision happens amongst boids we get the following exception,

```
Traceback (most recent call last):
  File "/home/kyber/Desktop/SMARTS/smarts/core/smarts.py", line 133, in step
    return self._step(agent_actions)
  File "/home/kyber/Desktop/SMARTS/smarts/core/smarts.py", line 202, in _step
    self._bubble_manager.step(self)
  File "/home/kyber/Desktop/SMARTS/smarts/core/bubble_manager.py", line 191, in step
    state_change = self.step_bubble_state(social_vehicles, social_agent_vehicles)
  File "/home/kyber/Desktop/SMARTS/smarts/core/bubble_manager.py", line 236, in step_bubble_state
    if not cursor.update():
  File "/home/kyber/Desktop/SMARTS/smarts/core/bubble_manager.py", line 113, in update
    pos = Point(self.vehicle.position)
  File "/home/kyber/Desktop/SMARTS/smarts/core/vehicle.py", line 215, in position
    pos, _ = self._chassis.pose.as_panda3d()
  File "/home/kyber/Desktop/SMARTS/smarts/core/chassis.py", line 357, in pose
    pos, orn = self._client.getBasePositionAndOrientation(self._bullet_id)
TypeError: an integer is required (got type NoneType)
```

This common behavior needs to be correctly handled.

**Comment by JVillella** [comment]
@Nickniyihan this is now working on master (not crashing anymore, however, the behavior is a bit complicated). Here's how it presently works,

* If any boid vehicle gets into a crash within the bubble zone the agent is killed and all vehicles taken over by SUMO.
* If any boid vehicle gets into a crash within the airlocking zone the agent is killed and then a new agent is immediately recreated (as a shadower) since one of the vehicles entered the airlocking zone. If this happens to happen within the final (exiting) airlock region the same happens, however we never end up hijacking and the shadowing agent gets terminated without ever having control.

I feel this behavior is a bit hard to follow, maybe even convoluted. But thinking through some of the alternatives, I see they have complicated edge cases as well. Does your team have a proposal on how you'd want it?

**Comment by Nickniyihan** [comment]
> @Nickniyihan this is now working on master (not crashing anymore, however, the behavior is a bit complicated). Here's how it presently works,
> 
> * If any boid vehicle gets into a crash within the bubble zone the agent is killed and all vehicles taken over by SUMO.
> * If any boid vehicle gets into a crash within the airlocking zone the agent is killed and then a new agent is immediately recreated (as a shadower) since one of the vehicles entered the airlocking zone. If this happens to happen within the final (exiting) airlock region the same happens, however, we never end up hijacking and the shadowing agent gets terminated without ever having control.
> 
> I feel this behavior is a bit hard to follow, maybe even convoluted. But thinking through some of the alternatives, I see they have complicated edge cases as well. Does your team have a proposal on how you'd want it?

Thank you @JVillella ! It is a bit complicated. To summarize, the cases can be:
1. Crash/offroad within airlock entering zone (where sumo controls vehicles)
2. Crash/offroad within bubble zone
3. Crash/offroad within airlock exiting zone (where agent still takes control)
And the ways to deal with these situations:
1. Replace dead bodies with sumo vehicles
2. Destroy the vehicle but keep the agent
3. Destroy the agent

I will communicate with Mohsen tomorrow.

**Comment by JVillella** [comment]
Chatted over zoom, closing for now.
